rand_distr/src/weighted_tree.rs

```diff
@@ -0,0 +1,381 @@
+/// selected element from the vector used to create the [`WeightedTreeIndex<W>`].
+/// The chance of a given element being picked is proportional to the value of
+/// the element. The weights can have any type `W` for which an implementation of
+/// [`Weight`] exists.
+///
+/// # Key differences
+///
+/// The main distinction between [`WeightedTreeIndex<W>`] and [`rand::distributions::WeightedIndex<W>`]
+/// lies in the internal representation of weights. In [`WeightedTreeIndex<W>`],
+/// weights are structured as a tree, which is optimized for frequent updates of the weights.
+///
+/// # Caution: Floating point types
+///
+/// When utilizing [`WeightedTreeIndex<W>`] with floating point types (such as f32 or f64),
+/// exercise caution due to the inherent nature of floating point arithmetic. Floating point types
+/// are susceptible to numerical rounding errors. Since operations on floating point weights are
+/// repeated numerous times, rounding errors can accumulate, potentially leading to noticeable
+/// deviations from the expected behavior.
+///
+/// Ideally, use fixed point or integer types whenever possible.
+///
+/// # Performance
+///
+/// A [`WeightedTreeIndex<W>`] with `n` elements requires `O(n)` memory.
+///
+/// Time complexity for the operations of a [`WeightedTreeIndex<W>`] are:
+/// * Constructing: Building the initial tree from an iterator of weights takes `O(n)` time.
+/// * Sampling: Choosing an index (traversing down the tree) requires `O(log n)` time.
+/// * Weight Update: Modifying a weight (traversing up the tree), requires `O(log n)` time.
+/// * Weight Addition (Pushing): Adding a new weight (traversing up the tree), requires `O(log n)` time.
+/// * Weight Removal (Popping): Removing a weight (traversing up the tree), requires `O(log n)` time.
+///
+/// # Example
+///
+/// ```
+/// use rand_distr::WeightedTreeIndex;
+/// use rand::prelude::*;
+///
+/// let choices = vec!['a', 'b', 'c'];
+/// let weights = vec![2, 0];
+/// let mut dist = WeightedTreeIndex::new(&weights).unwrap();
+/// dist.push(1).unwrap();
+/// dist.update(1, 1).unwrap();
+/// let mut rng = thread_rng();
+/// let mut samples = [0; 3];
+/// for _ in 0..100 {
+///     // 50% chance to print 'a', 25% chance to print 'b', 25% chance to print 'c'
+///     let i = dist.sample(&mut rng);
+///     samples[i] += 1;
+/// }
+/// println!("Results: {:?}", choices.iter().zip(samples.iter()).collect::<Vec<_>>());
+/// ```
+///
+/// [`WeightedTreeIndex<W>`]: WeightedTreeIndex
+#[cfg_attr(doc_cfg, doc(cfg(feature = "alloc")))]
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "serde1",
+    serde(bound(serialize = "W: Serialize, W::Sampler: Serialize"))
+)]
+#[cfg_attr(
+    feature = "serde1 ",
+    serde(bound(deserialize = "W: Deserialize<'de>, W::Sampler: Deserialize<'de>"))
+)]
+#[derive(Clone, Default, Debug, PartialEq)]
+pub struct WeightedTreeIndex<W: SampleUniform> {
```

Comment:
We require `W: Clone + PartialEq + PartialOrd + SampleUniform + SubAssign<W> + Weight` to sample which makes the distribution useless with any weaker bounds. It's not a set Rust style, but personally I am in favour of requiring all essential bounds on the type itself in cases like this.

(I am the author of the [`impl_scope!`](https://crates.io/crates/impl-tools#impl-scope) macro which makes repeating such bounds trivial. We don't really want that dependency here (it also has problems with `rustfmt` for now), but wish something like this were in Rust itself.)